### ISSUE TITLE
Просьба проверить тестовое задание

### ISSUE THREAD

**Issue by glebmlk** [glebmlk/wysiwig-task#2]
Привет, реализовал базовую функциональность используя `execCommand` в ветке https://github.com/glebmlk/wysiwig-task/tree/feat/execCommand. 
Описал несколько найденных багов в [issue](https://github.com/glebmlk/wysiwig-task/issues/1), в комментах к фиксам в коде есть на них ссылки. Багу №1 в Firefox починить не удалось.

Что изменил/добавил:
- Пришлось убрать `rem` единицы из `font-weight` и `line-height`, потому что не сохранялись стили текста при копировании и вставке в [MS Word Online](https://office.live.com/) (в Pages и Google Docs вставлялось нормально)
- XSS - вроде бы `contenteditable` сам разбирается со вставкой html-элементов и с обычным их вводом. Изучил несколько постов на [Stackoverflow](https://stackoverflow.com/questions/47995512/is-html5s-contenteditable-attribute-supposed-to-be-xss-safe), заглянул в OWASP и единственное, что я смог найти – прямая манипуляция с `innerHTML` у элемента с `contenteditable`. Написал проверку и вырезание всех не-`whitelisted` тегов и атрибутов в https://github.com/glebmlk/wysiwig-task/blob/feat/execCommand/js/xss-protection.js
В любом случае, необходимо добавлять проверки/санитизацию на абстрактном беке, так как можно подделать запрос или использовать другие способы атак.

Готов доделать/поправить комменты или переписать редактор с использованием, например, деревьев для более полноценной функциональности

**Comment by glebmlk** [comment]
Отдельно реализовал `wysiwyg` через `iframe` с `designMode = 'on'` в https://github.com/glebmlk/wysiwig-task/tree/feat/designMode

**Comment by JSMonk** [comment]
1. Можно посчитать текущий `font-weight` и `line-height` через `getComputedStyle`.
2. `innerHTML` - прямой путь к XSS, так что. постарайтесь обойтись без него.

Какую ветку по итогу проверять?

**Comment by glebmlk** [comment]
1. поправлю, добавлю в ветку https://github.com/glebmlk/wysiwig-task/tree/feat/execCommand
2. я как раз и не использую `innerHTML`, специально добавил проверку на изменение контента редактора через него (в devtools, например) в https://github.com/glebmlk/wysiwig-task/blob/feat/execCommand/js/xss-protection.js

проверять ветку https://github.com/glebmlk/wysiwig-task/tree/feat/execCommand, могу пингануть тогда, когда поправлю копипаст в Word Online

**Comment by JSMonk** [comment]
Хорошо. Можете сейчас просто заасайнить на Issue только себя, а потом, когда будет готово, меня.

**Comment by glebmlk** [comment]
можно смотреть ветку https://github.com/glebmlk/wysiwig-task/tree/feat/execCommand

поправил баг с `font-weight` в Firefox: по-умолчанию этот фикс выключен, чтобы включить надо выполнить `window.patchFirefox = true`

**Comment by JSMonk** [comment]
Здравствуйте.
Посмотрел ветку `feat/execCommand`:
у Вас не сохраняется форматироование, причем, если компировать то выглядит так:
<img width="657" alt="Screen Shot 2021-01-13 at 1 49 32 PM" src="https://user-images.githubusercontent.com/10776615/104448839-5b528980-55a6-11eb-9578-be04e682e328.png">
<img width="347" alt="Screen Shot 2021-01-13 at 1 49 39 PM" src="https://user-images.githubusercontent.com/10776615/104448843-5c83b680-55a6-11eb-864a-adca9a784083.png">

а если вырезать (CMD/CTRL + X), так:
<img width="657" alt="Screen Shot 2021-01-13 at 1 49 32 PM" src="https://user-images.githubusercontent.com/10776615/104448899-72917700-55a6-11eb-9f10-50f9779f4bf0.png">
<img width="229" alt="Screen Shot 2021-01-13 at 1 49 49 PM" src="https://user-images.githubusercontent.com/10776615/104448904-73c2a400-55a6-11eb-81ac-67ef9d41d1f9.png">

**Comment by glebmlk** [comment]
это при вставке в какой редактор?

**Comment by JSMonk** [comment]
В Microsoft Word Online

**Comment by glebmlk** [comment]
@JSMonk поправил сохранение форматирования в ветке `feat/execCommand`

**Comment by JSMonk** [comment]
Здравствуйте, Глеб.

При удалении и вставке содержимого теряется `font-weight`:

https://user-images.githubusercontent.com/10776615/104788972-5fe09300-579c-11eb-9788-e59242a45fe0.mov

**Comment by glebmlk** [comment]
есть еще другие баги, чтоб я сразу их пофиксил и не дергал вас постоянно?

**Comment by glebmlk** [comment]
этот баг поправил в `feat/execCommand`

**Comment by JSMonk** [comment]
Здравствуйте, Глеб.
Я показываю Вам все баги, которые нахожу.
1. Теряется перевод строки при вставке в Microsoft Office Online:
<img width="187" alt="Screen Shot 2021-01-17 at 6 02 29 PM" src="https://user-images.githubusercontent.com/10776615/104848566-3d5e8f00-58ee-11eb-8645-46a55479b951.png">
<img width="605" alt="Screen Shot 2021-01-17 at 6 02 23 PM" src="https://user-images.githubusercontent.com/10776615/104848567-3df72580-58ee-11eb-8673-7404ad3d0640.png">

2. Я не сразу заметил, но, попробуйте использовать уже заданные стили для заголовковв, italic и bold из `styles/editor.css`.

**Comment by glebmlk** [comment]
@JSMonk у Вас точно последняя версия ветки? у меня перенос строки сохраняется во всех браузерах

**Comment by glebmlk** [comment]
`использовать уже заданные стили для заголовков, italic и bold` имеете в виду вешать на элементы исходные классы?

**Comment by JSMonk** [comment]
Да, и придумать, как при копировании/вырезании сохранить стили для Microsoft Word Online.

**Comment by glebmlk** [comment]
- Вешать исходные классы обязательно? Я пытаюсь максимально делегировать применение стилей к тексту `execCommand` – навешивание классов будет излишним костылем (как мне кажется). Стили из `editor.css` применяются полноценно, просто по другим селекторам.
- Сохранение стилей поправил

**Comment by JSMonk** [comment]
Здравствуйте, Глеб.
Можно не навешивать классы, главное не меняйте относительные единицы на абсолютные.
Сегодня гляну.

**Comment by JSMonk** [comment]
Здравствуйте, Глеб.
Все хорошо, задание принято 👌

Можно Вашу почту (если подписаны были на Patreon - то ту же что и на патреоне) и telegram-контакт (номер или ник, все-равно) для дублирования сообщений.

Пройдет некоторое количество времени (мы сдвинем старт на неделю вперед) и Вы получите ссылку на оплату и мы сообщим Вам дальнейшие шаги.

**Comment by glebmlk** [comment]
Спасибо! Почта: glebmyalik@gmail.com, tg: `@glebmlk`

**Comment by xanf** [comment]
Обработано, сегодня получите письмо
